db/view/row_locking.cc

```diff
@@ -184,8 +184,27 @@ row_locker::unlock(const dht::decorated_key* pk, bool partition_exclusive,
             lock.read_unlock();
         }
         if (!lock.locked()) {
+            auto& row_locks = pli->second._row_locks;
+            // verify that all rows are unlocked.
+            // a row might be held by lock_ck if the lock_partition future timed out
+            // but the lock_row future succeeded,
+            // or there could be stale, unlocked rows, if the rwlock::holder
+            // was destroyed after the timeout before the lock_holder was created
+            // so unlock (this function) was never called.
```

Comment:
I don't understand this case exactly, but it also sounds like a bug, faulty exception handling (we want to create a lock_holder but before we do, we get an exception).

Reply:
Although the `rwlock::hold_{read|write)_lock` functions are not marked with noexcept, they actually are 
(and should be marked so), since all they so is call `get_units` that is noexcept.

Indeed, we will not create the lock_holder if any of the futures tracking the `rwlock::holder`s failed so we don't execute the `.then_unpack` continuation in `lock_ck`.  In this case, we may end up with a stale partition and/or row entries.